### DISCUSSION TITLE
Does the Redis client support the JSON.* commands?

### DISCUSSION THREAD

**Discussion by flowt-au** [quarkusio/quarkus#27610] [Q&A]
Hi. 
I am brand new to Quarkus and finding it wonderful. :-)

My initial use-case is to implement a Redis connection for the redisJSON and redisSearch modules. I have the demo app with basic key commands working using redis-stack in a Docker container. 

I can't find the class that exposes the JSON* commands. Does it exist?

If not, is there a workaround?

Thanks,
Murray

[a comment by an automated account is not shown]

**Comment by flowt-au** [comment]
I am guessing I might need to leverage the `ReactiveRedisDataSource.execute()`, something like below:
```
import io.quarkus.redis.datasource.ReactiveRedisDataSource;
...

public static Uni<Movie> findById(ReactiveRedisDataSource rrds, Long id) {
  // JSON.GET movie_json:1
  return rrds.execute("JSON.GET", "movie_json:" + id)
    .onItem()
    .doSomethingHere() // <== Reading up on what <Uni>Response actually means
  }
}
```

**Reply by cescoffier** [reply]
The JSON and Search commands are not part of the open-source Redis, so we do not have support for them for now (basically, I cannot test them). 

However, there are several approaches:

1. execute custom commands using the reactive or blocking data source
2. execute custom commands with the underlying Redis client

I believe the 1. is the simplest and is documented in https://quarkus.io/guides/redis-reference#executing-custom-commands. 

The object you receive is a `Response`, which is a container for the Redis command result. The `Response` lets you parse what Redis sent. I recommend checking the command documentation to understand the structure of the response.

**Reply by flowt-au** [reply]
Thanks Clement. Yes, I am on that track and reading the docs now. Thanks. Steep learning curve, but very excited. :-)

**Reply by flowt-au** [reply]
Ok. Finally solved it. But is this the best solution?

I wanted to send the Redis JSON response directly as a string, not via a bean, for now at least. So here I am converting the  `ReactiveRedisDataSource.execute()`  => `Uni<Response>` => `Uni<String>`

```java
public static Uni<String> findById(ReactiveRedisDataSource rrds, Long id) {
  
  // JSON.GET movie_json:1
  // returns eg {"id":1,"title":"Game of Thrones"}

  return rrds.execute("JSON.GET", "movie_json:" + id)
      .onItem()
      .transformToUni(resp -> MovieJSON(resp))
      .map(resp -> resp.toString());
}

private static Uni<String> MovieJSON(Response resp) {
  String respContent = (resp == null) ? "{}" : resp.toString();
  return Uni.createFrom().item(respContent);
}
```
Merci encore @cescoffier 

Murray
:-)

**Reply by cescoffier** [reply]
`resp.toString()` depends on the structure of the response sent by Redis. 
If I take the documentation of the "JSON.GET" command (from https://redis.io/commands/json.get/), I can see:

----
Return
[Array reply](https://redis.io/docs/reference/protocol-spec#resp-arrays) of [Bulk string reply](https://redis.io/docs/reference/protocol-spec#resp-bulk-strings) - each string is the JSON serialization of each JSON value that matches a path.

When using a JSONPath, the root of the matching values is always an array. In contrast, the legacy path returns a single value.

If there are multiple paths that include both legacy path and JSONPath, the returned value conforms to the JSONPath version (an array of values).
----

So, check if you don't have `[]` around the value. If you expect a single value, you can to `resp.get(0).toString()`

**Reply by flowt-au** [reply]
@cescoffier Thanks again, Clement. I am at a very early stage of learning with all this. At the start of the week I knew nothing about Redis, Quarkus or Mutiny, and my Java skills were very basic. I am saying this to thank you for taking the time to help while I get up to speed. I am gaining confidence that the Quarkus ecosystem is the right one for the fairy large project I am about to start, and a huge leap from the CFML world I was in until recently!

Again, much appreciated.
Murray

**Reply by flowt-au** [reply]
PS: I have to ask. Did the name Mutiny come from this TV Servies? [Halt and Catch Fire](https://en.wikipedia.org/wiki/Halt_and_Catch_Fire_(TV_series)). If not, and you haven't seen it, you might enjoy it! :-)
